[GOBBLIN-463] Change lineage event for Avro2Orc conversion to have underlying FileSystem as platform

aditya1105 opened the pull request:
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-463


### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
Changed HiveSource#setLineageInfo to create source and destination dataset descriptor using filesystem location. Changed ConvertibleHiveDatasetTest#testLineageInfo accordingly to accommodate the above change.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: ConvertibleHiveDatasetTest#testLineageInfo is the unit test.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/source/HiveSource.java:
```diff
@@ -495,11 +495,12 @@ private boolean isAvro(Table table) {
-    DatasetDescriptor source = new DatasetDescriptor(DatasetConstants.PLATFORM_HIVE, sourceTable);
-    source.addMetadata(DatasetConstants.FS_URI,
-        convertibleHiveDataset.getTable().getDataLocation().getFileSystem(new Configuration()).getUri().toString());
+    // Create source dataset descriptor
```
I understand the goal here is to provide hdfs based lineage for avro-to-orc conversion instead of hive based lineage. However, to keep the right semantic, `HiveSource` should indicate lineage from `Hive`.  We can do the following to achieve the goal:
- Add `sourceLocation` metadata to the existing hive lineage. `FS_URI` metadata isn't really required but more acts as an example of setting a metadata. Feel free to remove it
- Develop a avro-to-orc specific `DatasetResolver` to convert the hive based lineage to hdfs based.

zxcware commented on gobblin-api/src/main/java/org/apache/gobblin/dataset/DatasetConstants.java:
```diff
@@ -37,4 +37,8 @@
   public static final String CONNECTION_URL = "connectionUrl";
+
+  /** FileSystem schema and location */
+  public static final String FS_SCHEME = "fsPlatform";
```
to be consistent: `fsSchema`

zxcware commented on gobblin-api/src/main/java/org/apache/gobblin/dataset/HiveToHdfsDatasetResolver.java:
```diff
@@ -0,0 +1,50 @@
+  public DatasetDescriptor resolve(DatasetDescriptor raw, State state) {
+    ImmutableMap<String, String> metadata = raw.getMetadata();
+    Preconditions.checkArgument(metadata.containsKey(DatasetConstants.FS_SCHEME),
+        String.format("Hive Dataset Descriptor must contain property %s to create Hdfs Dataset Descriptor",
```
Let's be more specific: "must contain property" -> "must contain metadata"

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/source/HiveSource.java:
```diff
@@ -515,8 +517,10 @@ public static void setLineageInfo(ConvertibleHiveDataset convertibleHiveDataset,
       DatasetDescriptor dest = new DatasetDescriptor(DatasetConstants.PLATFORM_HIVE, destTable);
-      Path destPath = new Path(conversionConfigForFormat.get().getDestinationDataPath());
-      dest.addMetadata(DatasetConstants.FS_URI, destPath.getFileSystem(new Configuration()).getUri().toString());
+      FileSystem destFs = sourceFs;
```
Why is `destFs = sourceFs`? In general, a source can be ELT'ed to different destinations.

aditya1105 commented:
Update:
Removed changes from HiveSource for Lineage Event
Updated HiveAvroToOrcSource to add source lineage info
Updated HiveConvertPublisher to add destination lineage info
Updated HiveConvertibleDataset to contain source and dest DatasetDescriptors
Updated HiveConvertibleDatasetTest

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/dataset/ConvertibleHiveDataset.java:
```diff
@@ -112,6 +127,42 @@ public ConvertibleHiveDataset(FileSystem fs, HiveMetastoreClientPool clientPool,
+          .getDestinationTableName();
+      DatasetDescriptor dest = new DatasetDescriptor(DatasetConstants.PLATFORM_HIVE, destTable);
+      String destLocation = conversionConfigForFormat.get().getDestinationDataPath() + Path.SEPARATOR + "final";
+      // For Convertible HiveDataset source FileSystem will be equal to destination FileSystem
```
wrong comment? will be equal to source file system

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/source/HiveAvroToOrcSource.java:
```diff
@@ -36,8 +43,36 @@
   }
 
-}
+  public static boolean shouldSetLineageInfo(WorkUnit workUnit) {
```
How about `public static ConvertibleHiveDataset getConvertibleHiveDataset(WorkUnit workUnit)`, which returns the casted `convertibleHiveDataset` and `null` if not possible. Then we can have a cleaner logic
```
setSourceLineageInfo(workUnit, getConvertibleHiveDataset(workUnit), this.lineageInfo);
```
Remember to check null for `hiveDataset` in `setSourceLineageinfo`

eogren replied:
I actually like the shouldSetLineageInfo function - although I think, like your other comments, it should not be a static method just for unit test purposes

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/source/HiveAvroToOrcSource.java:
```diff
@@ -36,8 +43,36 @@
 
-    return super.getWorkunits(state);
+  @VisibleForTesting
+  public static void setSourceLineageInfo(WorkUnit workUnit, ConvertibleHiveDataset hiveDataset,
```
Remove the `static` so that subclass add its own information. For test, we can initialize a `HiveAvroToOrc` object.

zxcware commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/conversion/hive/publisher/HiveConvertPublisher.java:
```diff
@@ -253,6 +273,17 @@ public void publishData(Collection<? extends WorkUnitState> states) throws IOExc
   }
 
+  @VisibleForTesting
+  public static void setDestLineageInfo(WorkUnit workUnit, ConvertibleHiveDataset convertibleHiveDataset,
```
Remove `static` so that subclass can add its own information. For test, we can initialize a `HiveConvertPublisher` object.

eogren commented:
https://github.com/apache/incubator-gobblin/pull/2340 is a version updated with @zxcware's requested changes if it helps save some time

zxcware commented:
@aditya1105 We can close this pr, for Eric set up another one and merged your changes.
